### PR TITLE
New Drupal module for grouping menu links

### PR DESCRIPTION
**Pull request by Adnan-cds**
This module provides the following:
- A config entity that bundles several menu links under a group.
- A menu link plugin derivative to provide menu links for each group.
- A hook_menu_links_discovered_alter() implementation to move child menu links under their group's menu link.

@see https://github.com/localgovdrupal/localgov/issues/95

### CODE REVIEW

**Comment by Adnan-cds** [comment]
Please note that the README file of this module is not part this pull request.  I accidentally added it in the master branch :(

**Comment by andybroomfield** [comment]
Error on install
`
ParseError: syntax error, unexpected ')' in /app/web/modules/contrib/localgov_menu_link_group/src/Plugin/Deriver/MenuGroups.php on line 90 #0 /app/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/app/web/module...')
`

**Comment by Adnan-cds** [comment]
Ah Andy, this looks like a PHP version difference issue.  PHP 7.3 now allows [trailing commas in function calls](https://www.php.net/manual/en/migration73.new-features.php#migration73.new-features.core.trailing-commas).  I can think of two options here:
- Drop the trailing comma from the code involved.
- Declare PHP 7.3 as a requirement in the module info file.  Drupal 9's minimum [PHP requirement is 7.3](https://www.drupal.org/docs/understanding-drupal/how-drupal-9-was-made-and-what-is-included/environment-requirements-of#s-php-version-requirement)

Let me know what you would prefer. @andybroomfield

**Comment by andybroomfield** [comment]
@Adnan-cds this is one for Agile Collective (@finnlewis) as this is Lando environment that produces the error, and thats tagged to use PHP 7.2.

For now, I'd suggest removing it as technically the module is still Drupal 8 compatible and that can still run on PHP 7.2. (At least till November!)

**Comment by andybroomfield** [comment]
This works great from quick testing. Was able to create menu groups, and then export the configuration. 
Disabling the module, then enable and reimporting the config restores the menu groups.

Only issue aside from the install / php issue is to clear the menu cache after saving a group. This is so it doesn't have to be done manually each time.

Screenshot 2020-08-13 at 12 09 17 pm

**Comment by Adnan-cds** [comment]
> Only issue aside from the install / php issue

Sorry, I should have tested in Lando myself.  It should now work in PHP 7.2.

> is to clear the menu cache after saving a group. This is so it doesn't have to be done manually each time.

Yeah, annoying indeed.  I will see if I can find a work around.

**Comment by Adnan-cds** [comment]
Hi Andy,
I have added some code to rebuild the menu cache when Menu link groups are added/edited/deleted.  Let me know if that works.

FAO @andybroomfield

**Comment by andybroomfield** [comment]
This is working now, thanks @Adnan-cds.

I've discovered a potential bug, in that I accidently put a menu group into the user account menu and saved, then moved it into the correct place in add content and saved and it forgot which menu items where supposed to be in the group.

**Comment by Adnan-cds** [comment]
> it forgot which menu items where supposed to be in the group.

Thanks Andy :)  Yes, I can reproduce it.  Cross-menu assignment (e.g. admin menu to user menu) isn't working. I will add a test and a fix next week hopefully.

File: src/Entity/LocalGovMenuLinkGroup.php
Comment (stephen-cox) [inline src/Entity/LocalGovMenuLinkGroup.php]: These need doc blocks to say what they are. The following coding standards errors are thrown
```
----------------------------------------------------------------------
FOUND 5 ERRORS AFFECTING 5 LINES
----------------------------------------------------------------------
 52 | ERROR | Missing member variable doc comment
 54 | ERROR | Missing member variable doc comment
 56 | ERROR | Missing member variable doc comment
 58 | ERROR | Missing member variable doc comment
 60 | ERROR | Missing member variable doc comment
----------------------------------------------------------------------
```

**Review by stephen-cox — CHANGES_REQUESTED** [review]
Hi @Adnan-cds, Apologies in the delay in reviewing this. 

It looks to work well from my basic testing. There is a minor coding standards issue in that some class variables are missing docblocks, but nothing serious.

As the menu link groups are config entities is the idea that each module can provide config for the entities they provide? So the Campaigns module would provide something like:

```
langcode: en
status: true
dependencies: {  }
id: localgov_menu_link_group_campaigns
group_label: Campaign
weight: 0
parent_menu_link: 'admin:admin_toolbar_tools.extra_links:node.add'
child_menu_links:
  - 'admin:admin_toolbar_tools.extra_links:node.add.localgov_campaigns_overview'
  - 'admin:admin_toolbar_tools.extra_links:node.add.localgov_campaigns_page'
```

If so we need to think about dependencies. If the config is optional then it will we need think what happens if this module is enabled after modules that contain optional config. Ekes solved this by rescanning optional config on module install. See: https://github.com/localgovdrupal/localgov_step_by_step/blob/a32c9c7a1ebf02ffbc6444d0adda75e9e6106704/localgov_step_by_step.module#L181

The alternative would be for module to programmatically create the config on install. Something we've had to do for the Simple Sitemap module. See: https://github.com/localgovdrupal/localgov_step_by_step/blob/a32c9c7a1ebf02ffbc6444d0adda75e9e6106704/localgov_step_by_step.install#L11

File: src/Entity/LocalGovMenuLinkGroup.php
Comment (Adnan-cds) [reply src/Entity/LocalGovMenuLinkGroup.php]: Thanks Stephen, I will add the doc blocks.  It's a bit strange that my local copy of codesniffer doesn't complain about these.

As far as I am aware, optional configs are rescanned by Drupal during module installs.  I will double check this.

I realized last night that there may still be a bug in that multiple config files cannot share the same group.  For example: module A and module B may both want to their links in Group C, but that may not be possible with the current code.  I will try to fix this today.  I have just converted this pull request to draft.  Sorry for the chaos :(

Comment (Adnan-cds) [reply src/Entity/LocalGovMenuLinkGroup.php]: Hi Stephen,
I have added the doc blocks and fixed the bugs.  This is ready for review once again.

I have also checked the optional config issue.  I dropped a localgov_menu_link_group config entity file in the paragraphs module's config/optional/ directory and then turned on the localgov_menu_link_group module.  The menu changes applied fine.  So looks like Drupal is indeed [rescanning the optional config files during module installs](https://www.drupal.org/node/2453919)

Hi Andy,
The Cross menu bug should be fixed now.  Please give it a try.

FAO @stephen-cox @andybroomfield

**Comment by andybroomfield** [comment]
Hi @Adnan-cds Quick look, those fixes are working for me now (I can place menu group into a different menu and back and not lose the links). I'm happy with these changes and tests pass.

**Review by andybroomfield — APPROVED** [review]

**Comment by Adnan-cds** [comment]
Thanks Andy :)  Let's wait for Stephen.  If he's okay with the changes, we will merge it.

FAO @andybroomfield @stephen-cox

**Review by stephen-cox — APPROVED** [review]
Looks good to me - happy for this to be merged

**Comment by stephen-cox** [comment]
Once this has been merged it will need adding to [Packagist](https://packagist.org/packages/localgovdrupal/) and then as a Composer dependency in the [profile](https://github.com/localgovdrupal/localgov/blob/master/composer.json).

I have been doing the Packagist stuff and making Finn and Ekes maintainers. If @Adnan-cds  and @andybroomfield want to create accounts on Packagist I can make you maintainers of all the LocalGov packages as well.

**Comment by Adnan-cds** [comment]
> Looks good to me - happy for this to be merged

Thanks :)

> Once this has been merged it will need adding to Packagist and then as a Composer dependency in the profile.
> If @Adnan-cds and @andybroomfield want to create accounts on Packagist I can make you maintainers of all the LocalGov packages as well.

I will create a packagist account and pass you the details.  I will change the localgov profile once the module package is ready.  Hope that's okay.

FAO @stephen-cox

**Comment by andybroomfield** [comment]
@stephen-cox I should have a packagist account with my BHCC email address.

**Pull request merged**
